src/ext/random.c

```diff
@@ -19,32 +21,71 @@ void ddtrace_seed_prng(TSRMLS_D) {
 }
 
 void ddtrace_init_span_id_stack(TSRMLS_D) {
-    DDTRACE_G(root_span_id) = 0;
+    DDTRACE_G(trace_id) = 0;
     DDTRACE_G(span_ids_top) = NULL;
 }
 
 void ddtrace_free_span_id_stack(TSRMLS_D) {
+    DDTRACE_G(trace_id) = 0;
     while (DDTRACE_G(span_ids_top) != NULL) {
         ddtrace_span_ids_t *stack = DDTRACE_G(span_ids_top);
         DDTRACE_G(span_ids_top) = stack->next;
         efree(stack);
     }
 }
 
-uint64_t ddtrace_push_span_id(TSRMLS_D) {
+static inline uint64_t zval_to_uint64(zval *zid) {
+    if (!zid || Z_TYPE_P(zid) != IS_STRING) {
+        return 0U;
+    }
+    const char *id = Z_STRVAL_P(zid);
+#if PHP_VERSION_ID >= 70000
+    size_t i = 0;
+#else
+    int i = 0;
+#endif
+    for (; i < Z_STRLEN_P(zid); i++) {
+        if (id[i] < '0' || id[i] > '9') {
+            return 0U;
+        }
+    }
+    uint64_t uid = (uint64_t)strtoull(id, NULL, 10);
+    return (uid && errno != ERANGE) ? uid : 0U;
```

Comment:
You cannot inspect errno unless there was an error, so if there wasn't an error this code may misbehave.

Reply:
Nice catch!